x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/ServiceAccount.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc.service;
+
+import org.apache.logging.log4j.util.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.user.User;
+
+import java.io.IOException;
+import java.util.Objects;
+
+public interface ServiceAccount {
+
+    ServiceAccountId id();
+
+    RoleDescriptor roleDescriptor();
+
+    User asUser();
+
+    final class ServiceAccountId {
+
+        private final String namespace;
+        private final String serviceName;
+
+        public static ServiceAccountId fromPrincipal(String principal) {
+            final int split = principal.indexOf('/');
+            if (split == -1) {
+                throw new IllegalArgumentException(
+                    "a service account ID must be in the form {namespace}/{service-name}, but was [" + principal + "]");
+            }
+            return new ServiceAccountId(principal.substring(0, split), principal.substring(split + 1));
+        }
+
+        public ServiceAccountId(String namespace, String serviceName) {
+            this.namespace = namespace;
+            this.serviceName = serviceName;
+            validate();
+        }
+
+        public ServiceAccountId(StreamInput in) throws IOException {
+            this.namespace = in.readString();
+            this.serviceName = in.readString();
+        }
+
+        public void write(StreamOutput out) throws IOException {
+            out.writeString(namespace);
+            out.writeString(serviceName);
+        }
+
+        public String namespace() {
+            return namespace;
+        }
+
+        public String serviceName() {
+            return serviceName;
+        }
+
+        public String asPrincipal() {
+            return namespace + "/" + serviceName;
+        }
+
+        private void validate() {
```

Comment:
Is there a reason to split this into a separate method? It's only called from the constructor.